R/pkg/R/install.R

```diff
@@ -0,0 +1,155 @@
+  packageName <- ifelse(hadoop_version == "without",
+                        paste0(version, "-bin-without-hadoop"),
+                        paste0(version, "-bin-hadoop", hadoop_version))
+  if (is.null(local_dir)) {
+    local_dir <- getOption("spark.install.dir", spark_cache_path())
+  } else {
+    local_dir <- normalizePath(local_dir)
+  }
+
+  packageLocalDir <- file.path(local_dir, packageName)
+
+
+  # can use dir.exists(packageLocalDir) under R 3.2.0 or later
+  if (!is.na(file.info(packageLocalDir)$isdir)) {
+    fmt <- "Spark %s for Hadoop %s has been installed."
+    msg <- sprintf(fmt, version, hadoop_version)
+    message(msg)
+    return(invisible(packageLocalDir))
+  }
+
+  packageLocalPath <- paste0(packageLocalDir, ".tgz")
+  tarExists <- file.exists(packageLocalPath)
+
+  if (tarExists) {
+    message("Tar file found. Installing...")
+  } else {
+    if (is.null(mirror_url)) {
+      message("Remote URL not provided. Use Apache default.")
+      mirror_url <- mirror_url_default()
+    }
+
+    version <- "spark-2.0.0-rc4-bin"
+    # When 2.0 released, remove the above line and
+    # change spark-releases to spark in the statement below
+    packageRemotePath <- paste0(
+      file.path(mirror_url, "spark-releases", version, packageName), ".tgz")
+    fmt <- paste("Installing Spark %s for Hadoop %s.",
+                 "Downloading from:\n- %s",
+                 "Installing to:\n- %s", sep = "\n")
+    msg <- sprintf(fmt, version, hadoop_version, packageRemotePath,
+                   packageLocalDir)
+    message(msg)
+
+    fetchFail <- tryCatch(download.file(packageRemotePath, packageLocalPath),
+                          error = function(e) {
+                            msg <- paste0("Fetch failed from ", mirror_url, ".")
+                            message(msg)
+                            TRUE
+                          })
+    if (fetchFail) {
+      message("Try the backup option.")
+      mirror_sites <- tryCatch(read.csv(mirror_url_csv()),
+                               error = function(e) stop("No csv file found."))
+      mirror_url <- mirror_sites$url[1]
+      packageRemotePath <- paste0(file.path(mirror_url, version, packageName),
+                                  ".tgz")
+      message(sprintf("Downloading from:\n- %s", packageRemotePath))
+      tryCatch(download.file(packageRemotePath, packageLocalPath),
+               error = function(e) {
+                 stop("Download failed. Please provide a valid mirror_url.")
+               })
+    }
+  }
+
+  untar(tarfile = packageLocalPath, exdir = local_dir)
+  if (!tarExists) {
+    unlink(packageLocalPath)
+  }
+  message("Installation done.")
+  invisible(packageLocalDir)
+}
+
+mirror_url_default <- function() {
+  # change to http://www.apache.org/dyn/closer.lua
+  # when released
+
+  "http://people.apache.org/~pwendell"
+}
+
+supported_versions_hadoop <- function() {
+  c("2.7", "2.6", "2.4", "without")
+}
+
+spark_cache_path <- function() {
```

Comment:
Yes, those actually refer to the implementation of rappdirs/R/cache.r. Should add reference here.